R/check_tbl_value_col_ascending.R

```diff
@@ -70,9 +67,34 @@ check_tbl_value_col_ascending <- function(tbl, file_path, hub_path, round_id,
   )
 }
 
+#' Check that values for each model task in specific output types are ascending
+#'
+#' This function allows us to map over individual output types one at a time to
+#' reduce memory pressure.
+#' @param output_type the output type(s) to check. Must be a character vector
+#' @noRd
+check_values_ascending_by_output_type <- function(output_type, tbl,
+                                                  config_tasks, round_id,
+                                                  derived_task_ids) {
+  # FIX for <https://github.com/hubverse-org/hubValidations/issues/78>
+  # This function uses an inner join to auto-sort the table by model task,
+  # splitting by output type. We can use that to loop through the check.
+  output_type_tbls <- match_tbl_to_model_task(
+    tbl,
+    config_tasks = config_tasks,
+    round_id = round_id,
+    output_types = output_type,
+    derived_task_ids = derived_task_ids
+  ) %>%
+    purrr::compact()
+
+  purrr::map(output_type_tbls, check_values_ascending) %>%
+    purrr::list_rbind()
```

Comment:
Suggestion to make the variable name a bit more clear and remove the `map` and `list_rbind` that are not needed. 

```suggestion
  sorted_tbl <- match_tbl_to_model_task(
    tbl,
    config_tasks = config_tasks,
    round_id = round_id,
    output_types = output_type,
    derived_task_ids = derived_task_ids
  ) %>%
    purrr::compact() %>%
    purrr::pluck(1)
  check_values_ascending(sorted_tbl)
```

My reasoning: I didn't quite understand why `compact()`, `map()`, and `list_rbind()` were used here if we are testing a single output type. I went down a rabbit hole trying to figure out why I couldn't just pass the first item of output from `match_tbl_to_model_task()` to `check_values_ascending()` since we are subsetting by a single output type. But it turns out that the output of `match_tbl_to_model_task()` (and, subsequently, `expand_model_out_grid()`) returns one list item per output type.


Instead of using `map()` and `list_rbind()` to loop over a one-element list, I added `purrr::pluck(1)` to get the data frame and passed the value to `check_values_ascending()`.

Reply:
`match_tbl_to_model_task()` outputs **a list element per model task** not output type. If more than one output type is requested rows for all output types are included in each model task element.

We need compact to get rid of any `NULL` list elements which occur when an output type is not present in the config for a given model task. The code in this function is exactly as it appeared before in the main function, just wrapped in a new function so we can iterate over output types one at a type (which results in smaller outputs of `match_tbl_to_model_task()` at each iteration.

Therefore this suggestion won't work as you'd effectively only be processing the first model task.

Reply:
OHHHHHHH that's why! 

So the reason why the examples and tests only show one output type per element is because each model task has a separate output type? The reason it shows two elements is because there are two model tasks per round. 

If there were a hub that had two separate model tasks (e.g. for different targets) that were both providing quantile estimates, then `match_tbl_to_model_task()` would return two data frames when using `output_type = "quantile"`.

Reply:
Thank you for being patient with me on this.